### PR TITLE
python3packages.stytra: init at 0.8.26

### PR DESCRIPTION
**Pull request by tbenst**
WIP - wanted to get feedback on help as this is my first time packing using wheel.

It seems like `propogatedBuildInputs` is not behaving the same way--if I only have a package here I get failures like:

```Processing ./stytra-0.8.26-py3-none-any.whl
ERROR: Could not find a version that satisfies the requirement pyqtgraph>=0.10.0 (from stytra==0.8.26) (from versions: none)
ERROR: No matching distribution found for pyqtgraph>=0.10.0 (from stytra==0.8.26)
builder for '/nix/store/brscmbwx2nzr56f8wdh3n8vnjq2ih58p-python3.7-stytra-0.8.26.drv' failed with exit code 1
```
As a result, I'm including in both `buildInputs`, and `propogatedBuildInputs`, which feels wrong. It does build, but when I try open a shell the following happens:
```
> python
>>> import stytra
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/prr2jha833ix3qa30w0cbfgcr1zxh75s-python3-3.7.5-env/lib/python3.7/site-packages/stytra/__init__.py", line 3, in <module>
    from stytra.experiments import VisualExperiment
  File "/nix/store/prr2jha833ix3qa30w0cbfgcr1zxh75s-python3-3.7.5-env/lib/python3.7/site-packages/stytra/experiments/__init__.py", line 5, in <module>
    import numpy as np
ModuleNotFoundError: No module named 'numpy'
>>> import numpy
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'numpy'
```

This seems very odd, as in any other derivation I test, I can import from `propogatedBuildInputs` in python shell. To debug, I switched back to using the tar.gz file, and found an unexpected error:

```
these derivations will be built:
  /nix/store/25sy04zv5h91a02nkbdxy5lkf73ibn33-python3-3.7.5-env.drv
building '/nix/store/25sy04zv5h91a02nkbdxy5lkf73ibn33-python3-3.7.5-env.drv'...
collision between `/nix/store/10b0xgqaqacaxiwlvcklp9nbl20y5wcx-python3.7-sip-4.19.18/bin/sip' and `/nix/store/nqa4rbzhi6aj0ry54dwc82r2900zlnmr-python3.7-PyQt5.sip-4.19.18/bin/sip'
builder for '/nix/store/25sy04zv5h91a02nkbdxy5lkf73ibn33-python3-3.7.5-env.drv' failed with exit code 255
error: build of '/nix/store/25sy04zv5h91a02nkbdxy5lkf73ibn33-python3-3.7.5-env.drv' failed
```

Any thoughts or tips? Thanks!

###### Motivation for this change
Add module for closed-loop experiments for larval zebrafish.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/stytra/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/stytra/default.nix]: please add these as normal packages to python-modules, You should should have a commit for each one

File: pkgs/development/python-modules/stytra/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/stytra/default.nix]: you should prefer source over wheel... always
```suggestion
```

**Review by jonringer — COMMENTED** [review]
I don't see you declaring the pyqtgraph package in the propagated build inputs....

python packages should go into propagatedBuildInputs

File: pkgs/development/python-modules/stytra/default.nix
Comment (tbenst) [reply pkgs/development/python-modules/stytra/default.nix]: @FRidh's [post](https://github.com/NixOS/nixpkgs/pull/64047#issuecomment-522314445) made me think that building wheel is preferred?

**Comment by tbenst** [comment]
> I don't see you declaring the pyqtgraph package in the propagated build inputs....
> python packages should go into propagatedBuildInputs

It's there (via commonInputs). If pyqtgraph is only in propogatedBuildInputs, I get the error above. It goes away if I add to buildInputs, too, as is currently done in the code

File: pkgs/development/python-modules/stytra/default.nix
Comment (jonringer) [reply pkgs/development/python-modules/stytra/default.nix]: there's two things going on.... using a wheel from pypi, and building a wheel locally.

We prefer to receive source, and then as part of the buildPythonPackage workfow, build a wheel then install the wheel. This generally allows us to get tests, while only installing the files that package maintainer meant to distribute with a wheel.

We should prefer not to receive a wheel from pypi unless source isn't available from pypi or the upstream repo

Comment (tbenst) [reply pkgs/development/python-modules/stytra/default.nix]: ah, thanks for the clarification!

File: pkgs/development/python-modules/stytra/default.nix
Comment (tbenst) [reply pkgs/development/python-modules/stytra/default.nix]: done!

**Comment by tbenst** [comment]
This is blocked until https://github.com/NixOS/nixpkgs/pull/72378 is merged, but reviews in the meantime are welcome!

File: pkgs/development/python-modules/arrayqueues/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/arrayqueues/default.nix]: please don't use wheels, there's a source distribution
```suggestion
```

File: pkgs/development/python-modules/arrayqueues/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/arrayqueues/default.nix]:
```suggestion
```

File: pkgs/development/python-modules/colorspacious/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/colorspacious/default.nix]:
```suggestion
```

File: pkgs/development/python-modules/colorspacious/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/colorspacious/default.nix]:
```suggestion
```

**Review by jonringer — COMMENTED** [review]
please remove all instances of pulling down wheels when a source distribution is available, the only way a wheel should be allowed is if there no way to pull down source through pypi or the repo

**Comment by tbenst** [comment]
Thanks @jonringer. This is ready for final review

File: pkgs/development/python-modules/flammkuchen/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/flammkuchen/default.nix]:
```suggestion
  ] ++ lib.optionals isPy27 [ configparser ];
```

otherwise you get:
```
warning: ignoring the user-specified setting 'sandbox', because it is a restricted setting and you are not a trusted user
builder for '/nix/store/643f846aw8mdlblbx738am6hyhgayslc-python2.7-flammkuchen-0.9.1.drv' failed with exit code 1; last 10 log lines:
      from configparser import ConfigParser
  ImportError: No module named configparser


  ----------------------------------------------------------------------
  Ran 1 test in 0.000s

  FAILED (errors=1)
  Test failed: <unittest.runner.TextTestResult run=1 errors=1 failures=0>
  error: Test failed: <unittest.runner.TextTestResult run=1 errors=1 failures=0>
cannot build derivation '/nix/store/fqpy9j40cw5dn0m9sdclaricw7z9aj1n-env.drv': 1 dependencies couldn't be built
[17 built (1 failed), 19 copied (228.5 MiB), 45.0 MiB DL]
error: build of '/nix/store/fqpy9j40cw5dn0m9sdclaricw7z9aj1n-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/72353
1 package failed to build:
python27Packages.flammkuchen
```

**Review by jonringer — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/flammkuchen/default.nix
Comment (tbenst) [reply pkgs/development/python-modules/flammkuchen/default.nix]: thanks! committed

File: pkgs/development/python-modules/flammkuchen/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/flammkuchen/default.nix]:
```suggestion
{ lib, pkgs, buildPythonPackage, fetchPypi, isPy27
```

```
ache/nix-review/pr-72353/nixpkgs/lib/customisation.nix:121:8:
undefined variable 'isPy27' at /home/jon/.cache/nix-review/pr-72353/nixpkgs/pkgs/development/python-modules/flammkuchen/default.nix:27:22
```

Comment (tbenst) [reply pkgs/development/python-modules/flammkuchen/default.nix]: fixed

**Review by jonringer — CHANGES_REQUESTED** [review]
please squash the fixup to adhere to contributing guidelines

```
git reset HEAD^
git add pkgs/
git commit --amend --no-edit
git push <fork> <branch> --force
```

**Comment by tbenst** [comment]
@jonringer done

**Review by jonringer — APPROVED** [review]
diff LGTM
commits LGTM

```
https://github.com/NixOS/nixpkgs/pull/72353
12 package were built:
python27Packages.colorspacious python27Packages.flammkuchen python37Packages.arrayqueues python37Packages.colorspacious python37Packages.flammkuchen python37Packages.lightparam python37Packages.qimage2ndarray python37Packages.stytra python38Packages.arrayqueues python38Packages.colorspacious python38Packages.lightparam python38Packages.qimage2ndarray
```

**Pull request merged**
